docs/FileIO.md

```diff
@@ -0,0 +1,120 @@
+
+## The `FileIO` module {#sec-fileio}
+
+The `FileIO` module provides basic file I/O operations.
+Right now, these are limited to reading bytes from a file and writing bytes to a file.
+The API is intentionally limited in scope and will be expanded later.
+
+Unlike other modules in the `libraries` repository,
+the `FileIO` module will not compile or run correctly without a language-specific implementation file.
+Language-specific implementation files are provided for C#/.NET, Java, and Javascript.
+(Python and Golang support are planned.)
+Concretely, to use `FileIO` in your code, you must:
+
+1. `include` and `import` the `FileIO` module as you would any other library module
+2. incorporate the corresponding language-specific implementation file (e.g. `FileIO.cs`) when building or running your program
+
+For example, to run a `Program.dfy` file that depends on the `FileIO` module, run the following.
+(This assumes you have the `libraries` repository checked out within the working directory.)
+
+```bash
+# C#/.NET
+$ dafny run Program.dfy --include libraries/src/FileIO/FileIO.cs
+
+# Java
+$ dafny run Program.dfy --target:java --include libraries/src/FileIO/FileIO.java
+
+# Javascript
+$ dafny run Program.dfy --target:js --include libraries/src/FileIO/FileIO.js
+```
+
+(If you aren't using `dafny run` to run your program,
+then you should instead integrate the appropriate language-specific implementation file in your build system.)
+
+These sample programs from the examples folder illustrate how to read and write files:
+
+```dafny
+module ReadBytesFromFile {
+  import FileIO
+
+  method Main(args: seq<string>) {
+    expect |args| > 0;
+    expect |args| == 3, "usage: " + args[0] + " DATA_PATH EXPECTED_ERROR_PREFIX";
+    var dataPath := args[1];
+    var expectedErrorPrefix := args[2];
+
+    // Happy path: read from the data file, and check that we see the expected content.
+    {
+      var expectedStr := "Hello world\nGoodbye\n";
+      // This conversion is safe only for ASCII values. For Unicode conversions, see the Unicode modules.
+      var expectedBytes := seq(|expectedStr|, i requires 0 <= i < |expectedStr| => expectedStr[i] as int);
+
+      var res := FileIO.ReadBytesFromFile(dataPath);
+      expect res.Success?, "unexpected failure: " + res.error;
+
+      var readBytes := seq(|res.value|, i requires 0 <= i < |res.value| => res.value[i] as int);
+      expect readBytes == expectedBytes, "read unexpected byte sequence";
+    }
+
+    // Failure path: attempting to read from a blank file path should never work.
+    {
+      var res := FileIO.ReadBytesFromFile("");
+      expect res.Failure?, "unexpected success";
+      expect expectedErrorPrefix <= res.error, "unexpected error message: " + res.error;
```

Comment:
This is a good case for why strings are bad here.
Because a `FileNotFound` is not really an exceptional error case.
It should be easy to recover from.
But if you roll this library with string errors
it is complicated for customers to recover in this case.
This is especially hard in Dafny...

Reply:
Makes an easy example. But I thoroughly agree - location information and something better than a string as an error id are helpful for good error catching/reporting/recovery in even a moderate-sized system.